### PR TITLE
Adds Input Field Types mergeable validation rule

### PR DESCRIPTION
**Pull request by michaelstaib**

### CODE REVIEW

File: spec/Section 4 -- Composition.md [hunk 1]
```diff
@@ -14,6 +14,96 @@ run in sequence to produce the composite execution schema.
 
 ### Pre Merge Validation
 
+#### Input Field Types mergeable
+
+**Error Code**
+
+`INPUT_FIELD_TYPES_MERGABLE`
```
Comment (glen-84) [inline spec/Section 4 -- Composition.md:21]:
```suggestion
`INPUT_FIELD_TYPES_NOT_MERGEABLE`
```

File: spec/Section 4 -- Composition.md [hunk 2]
```diff
@@ -14,6 +14,96 @@ run in sequence to produce the composite execution schema.
...
+
+**Formal Specification**
+
+- Let {fieldsByName} be a map of field lists where the key is the name of a
+  field and the value is a list of fields from mergeable input types from
+  different source schemas with the same name.
+- For each {fields} in {fieldsByName}:
+  - if {InputFieldsAremergeable(fields)} must be true.
```
Comment (glen-84) [inline spec/Section 4 -- Composition.md:29]:
```suggestion
  - {InputFieldsAreMergeable(fields)} must be true.
```

File: spec/Section 4 -- Composition.md [hunk 3]
```diff
@@ -14,6 +14,96 @@ run in sequence to produce the composite execution schema.
...
+
+InputFieldsAremergeable(fields):
```
Comment (glen-84) [inline spec/Section 4 -- Composition.md:31]:
```suggestion
InputFieldsAreMergeable(fields):
```

File: spec/Section 4 -- Composition.md [hunk 4]
```diff
@@ -14,6 +14,96 @@ run in sequence to produce the composite execution schema.
...
+input Input1 {
+  field: String!
+}
+
+input Input1 {
+  field: String
+}
+```
+
+Fields are also consodered mergable if they have different nullability defined
```
Comment (glen-84) [inline spec/Section 4 -- Composition.md:63]:
```suggestion
Fields are also considered mergeable if they have different nullability defined.
```

File: spec/Section 4 -- Composition.md [hunk 5]
```diff
@@ -14,6 +14,96 @@ run in sequence to produce the composite execution schema.
...
+across Here, the field `tags` in `Input1` is a list type with compatible inner
```
Comment (glen-84) [inline spec/Section 4 -- Composition.md:64]:
```suggestion

Here, the field `tags` in `Input1` is a list type with compatible inner
```

File: spec/Section 4 -- Composition.md [hunk 6]
```diff
@@ -14,6 +14,96 @@ run in sequence to produce the composite execution schema.
...
+  tags: [String]!
+}
+
+input Input1 {
+  tags: [String]
+}
+```
+
+In this counter-example, the field `field` in `Input1` has incompatible types
+(`String` and `DateTime`), making them not mergeable:
```
Comment (glen-84) [inline spec/Section 4 -- Composition.md:82]:
```suggestion
(`String` and `DateTime`), making them non-mergeable:
```

**Review by glen-84 — CHANGES_REQUESTED** [review]

File: spec/Section 4 -- Composition.md [hunk 7]
```diff
@@ -14,6 +14,82 @@ run in sequence to produce the composite execution schema.
...
+  tags: [String]!
+}
+
+input Input1 {
+  tags: [String]
+}
+```
+
+In this example, the field `field` on `Input1` is not mergable as the field has
+different named types across (`String` and `DateTime`) source schemas:
```
Comment (glen-84) [inline spec/Section 4 -- Composition.md:81]:
```suggestion
In this example, the field `field` on `Input1` is not mergeable as the field has
different named types (`String` and `DateTime`) across source schemas:
```

**Comment by michaelstaib** [comment]
We need to explain how the nullability merges ...

File: spec/Section 4 -- Composition.md [hunk 8]
```diff
@@ -14,6 +14,82 @@ run in sequence to produce the composite execution schema.
...
+  different source schemas with the same name.
+- For each {fields} in {fieldsByName}:
+  - {InputFieldsAreMergeable(fields)} must be true.
+
+InputFieldsAreMergeable(fields):
+
+- Given each pair of members {fieldA} and {fieldB} in {fields}:
+  - Let {typeA} be the type of {fieldA}.
+  - Let {typeB} be the type of {fieldB}.
+  - {SameTypeShape(typeA, typeB)} must be true.
```
Comment (michaelstaib) [inline spec/Section 4 -- Composition.md:36]: this is wrong

Comment (PascalSenn) [reply spec/Section 4 -- Composition.md:36]: Resolved in 6564626

**Comment by PascalSenn** [comment]
Closed by #93

**Pull request closed without merging**